src/Hosting/Abstractions/src/BackgroundService.cs

```diff
@@ -29,11 +29,49 @@ public abstract class BackgroundService : IHostedService, IDisposable
         /// <param name="cancellationToken">Indicates that the start process has been aborted.</param>
         public virtual Task StartAsync(CancellationToken cancellationToken)
         {
-            // Create linked token to allow cancelling executing task from provided token
-            _stoppingCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            CancellationToken stoppingCtsToken;
 
+            // Only if _stoppingCts not created before or created and canceled
+            if (_stoppingCts?.IsCancellationRequested != false)
+            {
+                _stoppingCts = CancellationTokenSource.CreateLinkedTokenSource();
+
+                stoppingCtsToken = _stoppingCts.Token;
+
+                // Clear resources on cancel
+                stoppingCtsToken.Register(() =>
```

Comment:
It's not necessary to dispose a token that's already cancelled (or that's in the process of being cancelled), it just causes unnecessary ObjectDisposedExceptions for everyone.

Reply:
hmm ... in all my last code I relied on this article [https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.dispose?view=netcore-3.1](url):
_Always call Dispose before you release your last reference to the CancellationTokenSource. Otherwise, the resources it is using will not be freed until the garbage collector calls the CancellationTokenSource object's Finalize method._

Since CTS is never released in the original code, this could cause potential memory leaks.
Also in original code In the StartAsync() call CTS always rewriten with new instance (and new executedTask). And when you call StopAsync() you cancel only last CTS and last created executedTask.
Please correct me if I`m wrong...

Reply:
Calling Dispose is important if you're not going to call Cancel because yes it can leak memory (all the registered callbacks). Once cancelled, all the registered callbacks are released.

Reply:
I checked the sources of CTS, the Dispose() method clears other resources than Cancel().
But yes. I think you're right. This is not an infrastructure problem. Even if no one calls Dispose(), the GC can collect garbage when the CTS finalizer is called.

What do you think about:

> Also in original code In the StartAsync() call CTS always rewriten with new instance (and new executedTask). And when you call StopAsync() you cancel only last CTS and last created executedTask.

need to fix this?
maybe for BackgroundService.cs this is not such a big problem, since usually StartAsync() is called once.

Reply:
BackgroundService is not designed to be started more than once, so that's not a concern.

Reply:
ok. Last try. maybe rewrite Dispose() with 
`public virtual void Dispose()
        {
                using(_stoppingCts)
                {
                      _stoppingCts?.Cancel();
                }
        }`

Reply:
*shrug* not sure it accomplishes anything since the token gets cancelled anyways.